### PR TITLE
Adds option to customise error codes + useragent as a function

### PR DESCRIPTION
**Pull request by hasnat**
problem arises when using proxies, as most of them reply with 501 or 301/302 which we'd be considering as error codes since actual server is expect to send 2XX only

second thing is only changing option user agent to be either function or string (we're randomising user agents on every request)

### CODE REVIEW

File: lib/Request.js [hunk 1]
```diff
@@ -166,4 +165,22 @@ function extend(object, donor) {
...
+    error_status_codes = error_status_codes.split(',');
+    for (var index = 0, len = error_status_codes.length; index < len; ++index) {
+        var error_code = error_status_codes[index];
+        if (error_code.indexOf('-') > -1) {
+            var from_to = error_code.split('-').map(function (ec) { return parseInt(ec)});
+            if (response_error_code >= from_to[0] && response_error_code <= from_to[1]) {
+                return true;
+            }
+        }
+        if (response_error_code === parseInt(error_code)) {
```
Comment (rchipka) [inline lib/Request.js]: Add [base 10](https://stackoverflow.com/questions/6611824/why-do-we-need-to-use-radix)

File: index.js [hunk 2]
```diff
@@ -99,6 +99,7 @@ Osmosis.prototype.opts = {
     parse_cookies:          true, // Parse "Set-Cookie" header
     parse_response:         false,
     rejectUnauthorized:     false,
+    error_status_codes:     '400-500',
```
Comment (rchipka) [inline index.js]: Array support would be cool too. Not required.

Comment (rchipka) [reply index.js]: Better yet, a RegExp

File: lib/Request.js [hunk 3]
```diff
@@ -30,8 +30,7 @@ function Request(method, url, params, opts, tries, callback) {
 
         if (opts.ignore_http_errors !== true &&
             res                     !== undefined &&
-            res.statusCode          >=  400   &&
-            res.statusCode          <=  500) {
+            isErrorStatusCode(res.statusCode, opts.error_status_codes)) {
```
Comment (rchipka) [inline lib/Request.js]: Seems like a lot of string parsing to do on each request, perhaps we can pre-process to an array or better yet hash table lookup.

**Comment by rchipka** [comment]
@hasnat Please add tests for the user agent callback as well as the different error code input options.

Also, I'm not a big fan of the string parsing. Not huge on a RegExp either.

I vote for keeping things straight forward and accepting single int value, an array of integers, or a custom callback function that accepts the request's status code and returns a boolean value.

If you implement the callback, throw in the `res` response object as a second callback param for good  measure.

**Comment by hasnat** [comment]
thanks, I like the idea of callback with defaulting to 400-500 like it was before, as regexp for number range is bit awkward.
I'll update PR (with tests)

**Comment by hasnat** [comment]
@rchipka updated as per suggestions.

**Comment by hasnat** [comment]
I was thinking instead adding option `error_status_codes` and using `opts.error_status_codes(res.statusCode)`
how about I go with 
`is_expected_response` and use `opts.is_expected_response(res, data)`
this widens bit of checks you can do to determine if the response is good enough to continue or to be retried.

**Comment by rchipka** [comment]
@hasnat Great idea!

If we wanted to abstract that concept even further, we could have the existing `process_response` option callback accept `data`, `res`, and `next`.

If the given the callback accepts 3 arguments, then we only continue if it calls `next`.

Then you could check `res.statusCode` and the decision could be asynchronous, too.

Not required though, just a thought.

**Comment by rchipka** [comment]
Good job on the tests btw. I'll merge this in whenever you're ready.

**Comment by rchipka** [comment]
Ideally in the future we have [a better API for intercepting request information](https://github.com/rchipka/node-osmosis/issues/217). Added to [2.0 goals](https://github.com/rchipka/node-osmosis/projects/1).

**Comment by hasnat** [comment]
ah, I see `process_response`, do you mean something like this
https://github.com/rchipka/node-osmosis/compare/master...hasnat:patch-2 ?
this is nice but a breaking change

**Comment by rchipka** [comment]
@hasnat Yup. That change doesn't have to be breaking.

```javascript
if (typeof opts.process_response !== 'function') {
  if (opts.process_response.length > 2) {
    // async
  } else {
    // sync 
  }
}
```

**Comment by hasnat** [comment]
I see it won't be a breaking change if we do a count on arguments, but I'd like to send `callback` as argument too would be nice to call callback to say what the error was in `process_response`.

hold off merge then I'll remove `error_status_codes` and work on this.

Thanks

**Comment by hasnat** [comment]
@rchipka updated code to use `process_response`, many thanks for the directions.

**Comment by rchipka** [comment]
Publish v1.1.8

@hasnat Thanks and good work!

**Pull request merged**